cmd/habitat/community/state/transitions.go

```diff
@@ -378,3 +381,119 @@ func (t *StopProcessInstanceTransition) Validate(oldState *community.CommunitySt
 
 	return fmt.Errorf("process instance with matching process and node IDs not found")
 }
+
+type AddImplementationTransition struct {
+	InterfaceHash     string
+	DatastoreID       string
+	ContentIdentifier string
+}
+
+func (t *AddImplementationTransition) Type() string {
+	return TransitionTypeAddImplementation
+}
+
+func (t *AddImplementationTransition) Patch(oldState *community.CommunityState) ([]byte, error) {
+	// First try to see if there are other impls for the same hash that we can append to the list of
+	for _, impls := range oldState.DexImplementations {
+		if impls.InterfaceHash == t.InterfaceHash {
+			for _, impl := range impls.Implementations {
+				if impl.DatastoreID == t.DatastoreID {
+					return nil, fmt.Errorf("implementation with datastore %s already exists", t.DatastoreID)
+				}
+			}
+			return []byte(fmt.Sprintf(`[{
+				"op": "add",
+				"path": "/dex_implementations/%s/implementations/-",
+				"value": {
+					"datastore_id": "%s",
+					"content_identifier": "%s"
+				}
+			}]`, t.InterfaceHash, t.DatastoreID, t.ContentIdentifier)), nil
+		}
+	}
+
+	// If not available, create a new list
+	return []byte(fmt.Sprintf(`[{
+		"op": "add",
+		"path": "/dex_implementations/%s",
+		"value": {
+			"interface_hash": "%s",
+			"implementations": [
+				{
+					"datastore_id": "%s",
+					"content_identifier": "%s"
+				}
+			]
+		}
+	}]`, t.InterfaceHash, t.InterfaceHash, t.DatastoreID, t.ContentIdentifier)), nil
+}
+
+func (t *AddImplementationTransition) Validate(oldState *community.CommunityState) error {
+	// Check that datastore_id matches an existing process_id that has the is_datastore flag set, or that datastore_id is "ipfs"
+	validDatastore := false
+	if t.DatastoreID == "ipfs" {
+		validDatastore = true
+	}
+
+	for _, p := range oldState.Processes {
+		if p.ID == t.DatastoreID && p.IsDatastore {
```

Comment:
in the long term we don't want to use the process ID since this could be continually changing right?

Reply:
Good point. There are probably a couple ways we can handle this:
option 1: cascade changes to process state to all our impl data
option 2: create a consistent identifier for a datastore that persists across restarts

I think number 2 makes more sense. For now, I can leave a comment describing the problem and the options